In [1]:
import pandas as pd
import numpy as np

## Reading files

In [18]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [19]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [157]:
wh = pd.concat([wh2018,wh2019])

In [5]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [61]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [100]:
wh

Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.632           1.305   
1               2                    Norway  7.594           1.456   
2               3                   Denmark  7.555           1.351   
3               4                   Iceland  7.495           1.343   
4               5               Switzerland  7.487           1.420   
..            ...                       ...    ...             ...   
151           152                    Rwanda  3.334           0.359   
152           153                  Tanzania  3.231           0.476   
153           154               Afghanistan  3.203           0.350   
154           155  Central African Republic  3.083           0.026   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0             1.592                    0.874                         0.681   
1             1.582                    0.861                         0.686   
2             1.590                    0.868                         0.683   
3               NaN                    0.914                         0.677   
4             1.549                    0.927                         0.660   
..              ...                      ...                           ...   
151           0.711                    0.614                         0.555   
152           0.885                    0.499                         0.417   
153           0.517                    0.361                         0.000   
154           0.000                    0.105                         0.225   
155           0.575                    0.295                         0.010   

     Generosity  Perceptions of corruption  year  
0         0.202                      0.393  2018  
1         0.286                      0.340  2018  
2         0.284                      0.408  2018  
3           NaN                      0.138  2018  
4         0.256                      0.357  2018  
..          ...                        ...   ...  
151       0.217                      0.411  2019  
152       0.276                      0.147  2019  
153       0.158                      0.025  2019  
154       0.235                      0.035  2019  
155       0.202                      0.091  2019  

[312 rows x 10 columns]

In [54]:
wh.isna()

Overall rank  Country or region  Score  GDP per capita  Social support  \
0           False              False  False           False           False   
1           False              False  False           False           False   
2           False              False  False           False           False   
3           False              False  False           False            True   
4           False              False  False           False           False   
..            ...                ...    ...             ...             ...   
151         False              False  False           False           False   
152         False              False  False           False           False   
153         False              False  False           False           False   
154         False              False  False           False           False   
155         False              False  False           False           False   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
0                      False                         False       False   
1                      False                         False       False   
2                      False                         False       False   
3                      False                         False        True   
4                      False                         False       False   
..                       ...                           ...         ...   
151                    False                         False       False   
152                    False                         False       False   
153                    False                         False       False   
154                    False                         False       False   
155                    False                         False       False   

     Perceptions of corruption   year  
0                        False  False  
1                        False  False  
2                        False  False  
3                        False  False  
4                        False  False  
..                         ...    ...  
151                      False  False  
152                      False  False  
153                      False  False  
154                      False  False  
155                      False  False  

[312 rows x 10 columns]

In [55]:
wh.isna().sum()

Overall rank                     0
Country or region                0
Score                           19
GDP per capita                  12
Social support                  16
Healthy life expectancy         26
Freedom to make life choices    23
Generosity                      21
Perceptions of corruption       13
year                             0
dtype: int64

In [56]:
wh.isna().sum().sum()

130

In [57]:
x = wh[wh.isna().any(1)]

In [58]:
x

Overall rank Country or region  Score  GDP per capita  Social support  \
3               4           Iceland  7.495           1.343             NaN   
5               6       Netherlands    NaN           1.361           1.488   
7               8       New Zealand    NaN           1.268             NaN   
9              10         Australia  7.272           1.340           1.573   
12             13        Costa Rica  7.072           1.010             NaN   
..            ...               ...    ...             ...             ...   
137           138            Zambia    NaN           0.578           1.058   
140           141           Liberia    NaN           0.073           0.922   
142           143        Madagascar  3.933           0.274           0.916   
145           146          Zimbabwe  3.663           0.366           1.114   
150           151             Yemen    NaN           0.287           1.163   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
3                      0.914                         0.677         NaN   
5                      0.878                         0.638       0.333   
7                      0.876                         0.669       0.365   
9                      0.910                           NaN       0.361   
12                     0.817                         0.632       0.143   
..                       ...                           ...         ...   
137                    0.426                         0.431       0.247   
140                    0.443                           NaN       0.233   
142                    0.555                           NaN       0.169   
145                      NaN                         0.361       0.151   
150                    0.463                         0.143       0.108   

     Perceptions of corruption  year  
3                        0.138  2018  
5                        0.295  2018  
7                        0.389  2018  
9                        0.302  2018  
12                       0.101  2018  
..                         ...   ...  
137                        NaN  2019  
140                      0.033  2019  
142                      0.041  2019  
145                      0.089  2019  
150                      0.077  2019  

[107 rows x 10 columns]

In [78]:
x.groupby(['year']).sum(x.isna)

Overall rank    Score  GDP per capita  Social support  \
year                                                          
2018          4974  307.610          50.157          58.827   
2019          3524  165.279          35.488          52.587   

      Healthy life expectancy  Freedom to make life choices  Generosity  \
year                                                                      
2018                   31.065                        23.033       7.657   
2019                   20.886                        12.891       6.909   

      Perceptions of corruption  
year                             
2018                      5.610  
2019                      3.375

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [87]:
wh.describe()

Overall rank       Score  GDP per capita  Social support  \
count    312.000000  293.000000      300.000000      296.000000   
mean      78.500000    5.392816        0.896833        1.207145   
std       45.104737    1.108112        0.399329        0.302437   
min        1.000000    2.853000        0.000000        0.000000   
25%       39.750000    4.519000        0.588500        1.047750   
50%       78.500000    5.386000        0.954000        1.265500   
75%      117.250000    6.182000        1.221500        1.455500   
max      156.000000    7.769000        2.096000        1.624000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               286.000000                    289.000000  291.000000   
mean                  0.659832                      0.425872    0.183309   
std                   0.256373                      0.155103    0.096846   
min                   0.000000                      0.000000    0.000000   
25%                   0.485250                      0.324000    0.110000   
50%                   0.695500                      0.450000    0.175000   
75%                   0.859750                      0.543000    0.244000   
max                   1.141000                      0.724000    0.598000   

       Perceptions of corruption         year  
count                 299.000000   312.000000  
mean                    0.110950  2018.500000  
std                     0.095574     0.500803  
min                     0.000000  2018.000000  
25%                     0.050000  2018.000000  
50%                     0.082000  2018.500000  
75%                     0.139500  2019.000000  
max                     0.457000  2019.000000

In [83]:
data = wh.groupby('year')

In [86]:
data.describe()

Overall rank                                                    Score  \
            count  mean        std  min    25%   50%     75%    max  count   
year                                                                         
2018        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  151.0   
2019        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  142.0   

                ... Generosity        Perceptions of corruption            \
          mean  ...        75%    max                     count      mean   
year            ...                                                         
2018  5.354060  ...    0.23400  0.598                     149.0  0.112859   
2019  5.434028  ...    0.24825  0.566                     150.0  0.109053   

                                                     
           std  min      25%    50%      75%    max  
year                                                 
2018  0.098108  0.0  0.05000  0.082  0.13800  0.457  
2019  0.093279  0.0  0.04775  0.085  0.14025  0.453  

[2 rows x 64 columns]

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [148]:
wh

Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.632           1.305   
1               2                    Norway  7.594           1.456   
2               3                   Denmark  7.555           1.351   
3               4                   Iceland  7.495           1.343   
4               5               Switzerland  7.487           1.420   
..            ...                       ...    ...             ...   
151           152                    Rwanda  3.334           0.359   
152           153                  Tanzania  3.231           0.476   
153           154               Afghanistan  3.203           0.350   
154           155  Central African Republic  3.083           0.026   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0             1.592                    0.874                         0.681   
1             1.582                    0.861                         0.686   
2             1.590                    0.868                         0.683   
3               NaN                    0.914                         0.677   
4             1.549                    0.927                         0.660   
..              ...                      ...                           ...   
151           0.711                    0.614                         0.555   
152           0.885                    0.499                         0.417   
153           0.517                    0.361                         0.000   
154           0.000                    0.105                         0.225   
155           0.575                    0.295                         0.010   

     Generosity  Perceptions of corruption  year  
0         0.202                      0.393  2018  
1         0.286                      0.340  2018  
2         0.284                      0.408  2018  
3           NaN                      0.138  2018  
4         0.256                      0.357  2018  
..          ...                        ...   ...  
151       0.217                      0.411  2019  
152       0.276                      0.147  2019  
153       0.158                      0.025  2019  
154       0.235                      0.035  2019  
155       0.202                      0.091  2019  

[312 rows x 10 columns]

In [126]:
nan = wh.dropna(how='any')

In [158]:
wh.fillna(wh.groupby(['year']).transform('mean'), inplace=True)

C:\Users\RONYCA~1\AppData\Local\Temp/ipykernel_23544/2065541902.py:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.transform is deprecated. In a future version, a TypeError will be raised. Before calling .transform, select only columns which should be valid for the transforming function.
  wh.fillna(wh.groupby(['year']).transform('mean'), inplace=True)


In [159]:
wh

Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.632           1.305   
1               2                    Norway  7.594           1.456   
2               3                   Denmark  7.555           1.351   
3               4                   Iceland  7.495           1.343   
4               5               Switzerland  7.487           1.420   
..            ...                       ...    ...             ...   
151           152                    Rwanda  3.334           0.359   
152           153                  Tanzania  3.231           0.476   
153           154               Afghanistan  3.203           0.350   
154           155  Central African Republic  3.083           0.026   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0          1.592000                    0.874                         0.681   
1          1.582000                    0.861                         0.686   
2          1.590000                    0.868                         0.683   
3          1.205915                    0.914                         0.677   
4          1.549000                    0.927                         0.660   
..              ...                      ...                           ...   
151        0.711000                    0.614                         0.555   
152        0.885000                    0.499                         0.417   
153        0.517000                    0.361                         0.000   
154        0.000000                    0.105                         0.225   
155        0.575000                    0.295                         0.010   

     Generosity  Perceptions of corruption  year  
0       0.20200                      0.393  2018  
1       0.28600                      0.340  2018  
2       0.28400                      0.408  2018  
3       0.18136                      0.138  2018  
4       0.25600                      0.357  2018  
..          ...                        ...   ...  
151     0.21700                      0.411  2019  
152     0.27600                      0.147  2019  
153     0.15800                      0.025  2019  
154     0.23500                      0.035  2019  
155     0.20200                      0.091  2019  

[312 rows x 10 columns]

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [123]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro

In [166]:
wh.iloc[wh['Country or region'] == " " , :]

NotImplementedError: iLocation based boolean indexing on an integer type is not available